### PR TITLE
Fixes #39145 - Add Pulp 3.105 gems to gemspec; record VCR

### PR DESCRIPTION
**Pull request by qcjames53**
Related foreman-packaging PR #1: https://github.com/theforeman/foreman-packaging/pull/13391
Related foreman-packaging PR #2: https://github.com/theforeman/foreman-packaging/pull/13390

#### What are the changes introduced in this pull request?
- Updated plugin versions in Katello gemspec to match [Request for Pulpcore 3.105 Builds](https://community.theforeman.org/t/request-for-pulpcore-3-105-builds/45973) 
- Looked over monkey patches:
  - Updated `pulp_polymorphic_remote_response` to handle all viewsets instead of just a few.
  - Added `python_remote_provenance` patch to remove provenance field from PythonPythonRemote.
- Updated several Pulp async tasks to expect synchronous returns when using the new Pulp bindings.
- Full re-record of VCRs

#### Considerations taken when implementing this change?
Allegedly we will need to merge the two pulpcore-packaging PRs before this can hit Katello.

#### What are the testing steps for this pull request?
Ensure source looks correct and that full ktest passes in automation.

### CODE REVIEW

**Comment by qcjames53** [comment]
All VCR tests are passing so this should be ready for review! Gemspec changes are in https://github.com/Katello/katello/pull/11673

@sjha4  thank you for the monkey patch. You should be able to close out https://github.com/Katello/katello/pull/11714 if you'd like.

@ianballou Please let me know if you remember the secret extra VCRs which are missed by the regular commands. I'll add them to the PR.

**Comment by ianballou** [comment]
> @ianballou Please let me know if you remember the secret extra VCRs which are missed by the regular commands. I'll add them to the PR.

```bash
ktest test/actions/katello/repository/index_content_test.rb
ktest test/actions/katello/repository/multi_clone_contents_test.rb
ktest test/lib/tasks/pulpcore/repository_vcr_test.rb
ktest test/lib/util/errata_test.rb
ktest test/lib/util/pulpcore_content_filters_test.rb
ktest test/models/content_view_package_group_filter_test.rb
ktest test/models/katello/ostree_test.rb
ktest test/controllers/foreman/smart_proxies_controller_test.rb
```

I think these are all of them, there are actually more than I remember.

File: app/services/katello/pulp3/repository.rb [hunk 1]
```diff
@@ -324,7 +329,10 @@ def update_distribution
           end
           content_guard_prn = options.delete(:content_guard_prn) # Extract PRN and remove from options
           distribution_reference.update(:content_guard_href => options[:content_guard], :content_guard_prn => content_guard_prn)
-          api.distributions_api.partial_update(distribution_reference.href, options)
+          response = api.distributions_api.partial_update(distribution_reference.href, options)
```
Comment (ianballou) [inline app/services/katello/pulp3/repository.rb:331]: I want to locate the Pulp change for this. We need to very carefully check all API changes, since so often we find issues in the Ruby bindings. I've only found the remote one: https://github.com/pulp/pulpcore/issues/6897

File: app/services/katello/pulp3/repository.rb
Comment (ianballou) [inline app/services/katello/pulp3/repository.rb]: We need to test, but from the Pulp team it sounds like all sorts of update methods now return nil or a task. Repository update is one major one I suspect.

File: test/actions/pulp3/orchestration/yum_refresh_if_needed_test.rb [hunk 2]
```diff
@@ -19,7 +19,9 @@ def teardown
 
     def test_refresh_if_needed
       action = ForemanTasks.sync_task(::Actions::Pulp3::Orchestration::Repository::RefreshIfNeeded, @repo, @smart_proxy)
-      refute_empty action.output[:pulp_tasks]
+      # In Pulp 3.105+, if there are no changes needed, no tasks are created (200 OK instead of 202 Accepted)
+      # The action should complete successfully even with no tasks
+      assert action.output[:pulp_tasks].is_a?(Array)
```
Comment (ianballou) [inline test/actions/pulp3/orchestration/yum_refresh_if_needed_test.rb]: Worth adding an empty array assertion?

**Review by ianballou — COMMENTED** [review]
First pass review, biggest thing is that the Pulp team mentioned all (most?) update tasks now return a polymorphic response.
Thanks for getting through all the VCRs!

File: app/services/katello/pulp3/repository.rb
Comment (ianballou) [reply app/services/katello/pulp3/repository.rb]: ### All affected models (via `AsyncUpdateMixin`)

| Pulpcore ViewSet                    | Affected Resources                                                                              |
|-------------------------------------|-------------------------------------------------------------------------------------------------|
| **`RemoteViewSet`**                 | All remotes (RPM, ULN, File, Ansible Collection/Git/Role, Container, PullThrough, Deb, Python, OSTree)   |
| **`RepositoryViewSet`**             | All repositories (RPM, File, Ansible, Container, Deb, Python, OSTree)                           |
| **`DistributionViewSet`**           | All distributions (RPM, File, Ansible, Container, PullThrough, Deb, Python, OSTree)             |
| **`ExporterViewSet`**               | Pulp exporters                                                                                  |
| **`AlternateContentSourceViewSet`** | RPM ACS, Deb ACS                                                                                |
| **`DomainViewSet`**                 | Pulp domains                                                                                    |

File: test/actions/pulp3/orchestration/yum_refresh_if_needed_test.rb (hunk 2 above)
Comment (qcjames53) [reply test/actions/pulp3/orchestration/yum_refresh_if_needed_test.rb]: I can add that.

File: app/services/katello/pulp3/repository.rb
Comment (qcjames53) [reply app/services/katello/pulp3/repository.rb]: I *think* the newest push should address all of these resources. This ended up being a two part fix requiring binding mokney patches (looking pretty snazzy now) and Katello-side pulp3 module changes.

File: app/services/katello/pulp3/repository.rb (hunk 1 above)
Comment (qcjames53) [reply app/services/katello/pulp3/repository.rb:331]: I was able to confirm that the distribution api's `partial_update` method has the exact same polymorphic behavior as the other updates since it inherits `AsyncUpdateMixin` in Pulp. It returns 202 when changes occur and 200 when no changes are detected.

**Comment by qcjames53** [comment]
I think this is ready for review. I'll create a redmine and fix the PR description tomorrow morning.

File: lib/monkeys/python_remote_provenance.rb [hunk 3]
```diff
@@ -0,0 +1,163 @@
+# Monkey-patch PythonPythonRemote to skip provenance field for backwards compatibility
```
Comment (ianballou) [inline lib/monkeys/python_remote_provenance.rb:1]: This monkey patching is getting intense, now hopefully I have another strong ally for getting us off of the generated Pulp bindings :)

**Comment by qcjames53** [comment]
This new push should get rid of a few of the failures. We'll have maybe 5-10 failures left fingers crossed.

**Comment by qcjames53** [comment]
@sjha4 fixed your email address :)

**Comment by ianballou** [comment]
I pushed a change that should get the scenario tests working again.

File: lib/monkeys/python_remote_provenance.rb [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    end
+
+    if attributes.key?(:'max_retries')
+      self.max_retries = attributes[:'max_retries']
+    end
+
+    if attributes.key?(:'policy')
+      self.policy = attributes[:'policy']
+    else
+      self.policy = 'immediate'
```
Comment (ianballou) [inline lib/monkeys/python_remote_provenance.rb]: The only anomaly I see here is that `immediate` is a default. It should be totally safe and it makes sense, but I am curious why it's different from what the Python 3.27.2 gem has:

```
      if attributes.key?(:'policy')
        self.policy = attributes[:'policy']
      end

```

**Review by ianballou — APPROVED** [review]
ACK with one question, thanks!

I double checked that we see task details for publications, distributions, remotes, etc.

File: lib/monkeys/python_remote_provenance.rb (hunk 4 above)
Comment (qcjames53) [reply lib/monkeys/python_remote_provenance.rb]: I guarantee this is a Claude add. Although this section is from Samir, I had a bunch of experience with Claude seeming to think immediate vs on demand was a big deal for Python packages when the field really DOES NOT matter. I did think about removing this but left it in since Claude would just keep trying to put it back :)

Comment (qcjames53) [reply lib/monkeys/python_remote_provenance.rb]: We can remove it if you'd like. I meant to delete it tbh.

Comment (ianballou) [reply lib/monkeys/python_remote_provenance.rb]: I didn't realize the default policy should actually be on_demand, so +1 to removing it https://pulpproject.org/pulp_python/restapi/#tag/Remotes:-Python/operation/remotes_python_python_create

**Review by ianballou — CHANGES_REQUESTED** [review]

**Comment by qcjames53** [comment]
This most recent push should address that concern. I needed to re-record a few tests. Fingers crossed that CI passes!

**Comment by ianballou** [comment]
CI failure was a transient issue - rerunning.

**Review by ianballou — APPROVED** [review]
Thanks!

**Review by sjha4 — APPROVED** [review]
Tests green..Changes LGTM..Packaging PRs are green..Let's start merging things.. 🎉
